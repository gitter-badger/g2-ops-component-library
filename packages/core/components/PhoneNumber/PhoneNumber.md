Phone Number Example

```js
<div style={{ maxWidth: '300px' }}>
  <PhoneNumber label="Enter Phone Number" value="4934382481830" countryCode="de" />
</div>
```
