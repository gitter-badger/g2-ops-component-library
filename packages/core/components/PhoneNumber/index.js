export { AsYouType, PhoneNumber } from './PhoneNumber'
