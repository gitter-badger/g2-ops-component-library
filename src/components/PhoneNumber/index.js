export { AsYouType, PhoneNumber } from './PhoneNumber'
export default from './PhoneNumber'
