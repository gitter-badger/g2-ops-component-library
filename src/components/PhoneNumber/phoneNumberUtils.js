import countryPhoneData from 'country-telephone-data'
import { pathOr } from 'ramda'
import { AsYouType } from 'libphonenumber-js'

export const getCountryDialCode = (countryCode = '') => {
  const countryPhoneDataIndex = pathOr('', [ 'iso2Lookup', countryCode.toLowerCase() ], countryPhoneData)
  return pathOr('', [ 'allCountries', countryPhoneDataIndex, 'dialCode' ], countryPhoneData)
}

export const stripDownPhoneNumber = (num = '') => num.replace(/\D+/g, '')

export const getFormattedNumber = (value: string = ''): string => {
  const unformattedValue = `+${stripDownPhoneNumber(value)}`
  if (value.length <= 2) {
    return unformattedValue
  } else {
    const asYouTypeFormatter = new AsYouType()
    const formattedNumber = asYouTypeFormatter.input(unformattedValue)
    return formattedNumber
  }
}
