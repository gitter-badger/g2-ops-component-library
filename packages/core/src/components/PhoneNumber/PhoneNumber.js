// @flow

import React, { Component } from 'react'
import { pathOr, isEmpty } from 'ramda'
import { TextField } from 'components/TextField'
import { AsYouType } from 'libphonenumber-js'
import { stripDownPhoneNumber, getCountryDialCode, getFormattedNumber } from './phoneNumberUtils'

type PhoneNumberPropsT = {
  value: string,
  label: string,
  placeholder: string,
  countryCode: string,
}

type PhoneNumberStateT = {
  formattedNumber: string,
}

export class PhoneNumber extends Component<PhoneNumberPropsT, PhoneNumberStateT> {
  state = {
    formattedNumber: getFormattedNumber(this.props.value),
  }

  numberInput: Node
  handleChange = (value: string) => {
    const formattedNumber: string = getFormattedNumber(value)
    let caretPosition = this.numberInput.selectionStart
    const oldFormattedText = this.state.formattedNumber
    const diff = formattedNumber.length - oldFormattedText.length

    const onSetStateComplete = () => {
      if (caretPosition === 1 && formattedNumber.length === 2) {
        caretPosition += 1
      }
      if (diff > 0) {
        caretPosition -= diff
      }
      if (caretPosition > 0 && oldFormattedText.length >= formattedNumber.length) {
        this.numberInput.setSelectionRange(caretPosition, caretPosition)
      }
    }
    this.setState({ formattedNumber }, onSetStateComplete)
  }

  handleFocus = () => {
    const unformattedPhoneNumber = stripDownPhoneNumber(this.state.formattedNumber)
    if (isEmpty(unformattedPhoneNumber)) {
      this.handleChange(getCountryDialCode(this.props.countryCode))
    }
  }
  render() {
    const { value, placeholder, label, ...otherProps } = this.props
    return (
      <TextField
        {...otherProps}
        label={label}
        componentRef={input => {
          this.numberInput = input
        }}
        onChanged={this.handleChange}
        value={this.state.formattedNumber}
        placeholder={placeholder || label}
        onFocus={this.handleFocus}
      />
    )
  }
}

export { AsYouType }
