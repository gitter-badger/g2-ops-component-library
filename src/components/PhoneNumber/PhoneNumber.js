// @flow

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { pathOr, isEmpty } from 'ramda'
import TextField from 'components/TextField'
import { AsYouType } from 'libphonenumber-js'
import { stripDownPhoneNumber, getCountryDialCode, getFormattedNumber } from './phoneNumberUtils'

type PhoneNumberPropTypes = {
  value: string,
  label: string,
  placeholder: string,
  countryCode: string,
}

type PhoneNumberStateTypes = {
  formattedNumber: string,
}

class PhoneNumber extends Component<PhoneNumberPropTypes, PhoneNumberStateTypes> {
  state = {
    formattedNumber: getFormattedNumber(this.props.value),
  }

  numberInput: Node
  handleChange = (value: string) => {
    const formattedNumber: string = getFormattedNumber(value)
    let caretPosition = this.numberInput.selectionStart
    const oldFormattedText = this.state.formattedNumber
    const diff = formattedNumber.length - oldFormattedText.length

    const onSetStateComplete = () => {
      if (caretPosition === 1 && formattedNumber.length === 2) {
        caretPosition += 1
      }
      if (diff > 0) {
        caretPosition -= diff
      }
      if (caretPosition > 0 && oldFormattedText.length >= formattedNumber.length) {
        this.numberInput.setSelectionRange(caretPosition, caretPosition)
      }
    }
    this.setState({ formattedNumber }, onSetStateComplete)
  }

  handleFocus = () => {
    const unformattedPhoneNumber = stripDownPhoneNumber(this.state.formattedNumber)
    if (isEmpty(unformattedPhoneNumber)) {
      this.handleChange(getCountryDialCode(this.props.countryCode))
    }
  }
  render() {
    const { value, placeholder, label, ...otherProps } = this.props
    return (
      <TextField
        {...otherProps}
        label={label}
        componentRef={(input) => {
          this.numberInput = input
        }}
        onChanged={this.handleChange}
        value={this.state.formattedNumber}
        placeholder={placeholder || label}
        onFocus={this.handleFocus}
      />
    )
  }
}

export { AsYouType }
export default PhoneNumber
