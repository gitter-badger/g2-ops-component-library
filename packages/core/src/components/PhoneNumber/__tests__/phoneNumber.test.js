import React from 'react'
import { mount } from 'enzyme'
import {PhoneNumber} from '../PhoneNumber'
import toJSON from 'enzyme-to-json'

describe('<PhoneNumber />', () => {
  test('should render properly for various country codes', () => {
    const countries = [ 'US', 'UK', 'IR', 'ME', 'CA', 'DE', 'ES', 'IN', 'GB' ]
    countries.forEach((countryCode) => {
      const tree = mount(<PhoneNumber label={'Phone Field'} disabled={false} countryCode={countryCode} />)
      expect(toJSON(tree)).toMatchSnapshot()
    })
  })

  test('should render properly with value set', () => {
    const tree = mount(
      <PhoneNumber label={'Phone Field'} disabled={false} countryCode={'de'} value={'779384741'} />
    )
    expect(toJSON(tree)).toMatchSnapshot()
  })
})
