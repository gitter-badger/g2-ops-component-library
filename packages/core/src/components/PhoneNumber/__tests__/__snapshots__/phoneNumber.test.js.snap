// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 1`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField0"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField0"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 2`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField2"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField2"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 3`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField4"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField4"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 4`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField6"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField6"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 5`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField8"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField8"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 6`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField10"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField10"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 7`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField12"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField12"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 8`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField14"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField14"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly for various country codes: [object Object] 9`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField16"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField16"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<PhoneNumber /> should render properly with value set: [object Object] 1`] = `
<div
  className="ms-Fabric root-33"
>
  <div
    className="ms-TextField root"
  >
    <div
      className="ms-TextField-wrapper wrapper"
    >
      <label
        className="ms-Label root-34"
        htmlFor="TextField18"
      >
        Phone Field
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup"
      >
        <input
          aria-describedby={undefined}
          aria-invalid={false}
          aria-label={undefined}
          className="ms-TextField-field field input "
          disabled={false}
          id="TextField18"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          placeholder="Phone Field"
          type="text"
          value="+7 793 847 41"
        />
      </div>
    </div>
  </div>
</div>
`;
